contributors/design-proposals/cloud-provider-refactoring.md

```diff
@@ -0,0 +1,166 @@
+
+The routeController configures routes for hosts in Google Cloud.
+
+The serviceController maintains a list of currently active nodes, and is responsible for creating and deleting LoadBalancers in the underlying cloud.
+
+#### Cloud Provider in Kubelet
+
+Moving on to the kubelet, the following cloud provider dependencies exist in kubelet.
+
+ - Find the cloud nodename of the host that kubelet is running on for the following reasons :
+      1. To obtain the config map for the kubelet, if one already exists
+      2. To uniquely identify current node using nodeInformer
+      3. To instantiate a reference to the current node object
+ - Find the InstanceID, ProviderID, ExternalID, Zone Info of the node object while initializing it
+ - Periodically poll the cloud provider to figure out if the node has any new IP addresses associated with it
+ - In case of Google cloud, it allows the provider to configure routes for the node
+ - It allows the cloud provider to post process DNS settings
+
+#### Cloud Provider in Kube-apiserver
+
+Finally, in the kube-apiserver, the cloud provider is used for transferring SSH keys to all of the nodes, and within an admission controller for setting labels on persistent volumes.
+
+### 2. Strategy for refactoring Kube-Controller-Manager
+
+In order to create a 100% cloud independent controller manager, we will split the controller-manager into multiple binaries.
+
+1. Cloud dependent controller-manager binaries
+2. Cloud independent controller-manager binaries - This is the existing `controller-manager` thats shipped with kubernetes releases presently.
+
+The cloud dependent binaries will run those loops that rely on cloudprovider as a kubernetes system service (started using `kubectl apply -f`). The rest of the controllers will be run in the cloud independent controller manager. The decision to run entire controller loops, rather than only the very minute parts that rely on cloud provider makes implementation simple. If drilled down further into the controller loops, it would be significantly harder to disentangle, due to tight coupling of the cloud provider specific code with the rest of the data structures/functions within a controller. There would be a lot of duplicated code, which needs to be kept in sync, across the "dependent" loop implementations. For example, StatefulSet would not have gotten to beta in 1.5 if we had to change 7 node controller loops in various cloud provider binaries.
+
+Note that the controller loop implementation will continue to reside in the core repository. It takes in cloudprovider.Interface as an input in its constructor. Vendor maintained cloud-controller-manager binary will link these controllers in. It will also initialize the cloudprovider object and pass it into the controller loops.
```

Comment:
@mfischer-zd It might help to look up how the informer framework works in kubernetes. This is my favourite document about the informer framework - http://jayunit100.blogspot.com/2015/10/kubernetes-informers-controllers.html

> Things I'm interested in include:

> "reacting to cloud events" - how? What's the API, if any?

Reading up about informer framework is essential to understanding this. There is no cloud API per se. This is an example flow - When a kubelet registers a node with the apiserver, the controller manager reacts to this creation by talking to the cloudprovider and fetching information about the node. The information about the creation is brought to the controller manager via the informers. 

> How to ensure this custom cloud-controller-manager would be a singleton to avoid duplicate and possibly conflicting event consumption/reaction

First of all, multiple custom controller managers can be simultaneously with leader election. Secondly, Kubernetes objects are idempotent by design. Two objects will not step on each other because of this.